src/bridge/SequencerInbox.sol

```diff
@@ -307,37 +322,60 @@ contract SequencerInbox is GasRefundEnabled, ISequencerInbox {
         return (header, timeBounds);
     }
 
-    function formDataHash(bytes calldata data, uint256 afterDelayedMessagesRead)
+    function formEmptyDataHash(uint256 afterDelayedMessagesRead)
         internal
         view
-        validateBatchData(data)
         returns (bytes32, IBridge.TimeBounds memory)
     {
         (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
             afterDelayedMessagesRead
         );
-        bytes32 dataHash = keccak256(bytes.concat(header, data));
-        return (dataHash, timeBounds);
+        return (keccak256(header), timeBounds);
     }
 
-    function formEmptyDataHash(uint256 afterDelayedMessagesRead)
-        internal
-        view
-        returns (bytes32, IBridge.TimeBounds memory)
-    {
+    function formDataHash(
+        bytes calldata data,
+        uint256 afterDelayedMessagesRead,
+        IBridge.BatchDataLocation dataLocation
+    ) internal view returns (bytes32, IBridge.TimeBounds memory) {
+        uint256 fullDataLen = HEADER_LENGTH + data.length;
+        if (fullDataLen > maxDataSize) revert DataTooLarge(fullDataLen, maxDataSize);
+        if (data.length > 0 && (data[0] & DATA_AUTHENTICATED_FLAG) == DATA_AUTHENTICATED_FLAG) {
+            revert DataNotAuthenticated();
+        }
+
         (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
             afterDelayedMessagesRead
         );
-        return (keccak256(header), timeBounds);
+        if (dataLocation == IBridge.BatchDataLocation.Blob) {
+            bytes32[] memory dataHashes = dataHashReader.getDataHashes();
+            if (dataHashes.length == 0) revert MissingDataHashes();
+            if (data.length != 1) revert InvalidBlobMetadata();
+            if (data[0] & DATA_BLOB_HEADER_FLAG == 0) revert InvalidBlobMetadata();
+
+            return (
+                keccak256(bytes.concat(header, data, abi.encodePacked(dataHashes))),
+                timeBounds
+            );
+        } else {
+            // the first byte is used to identify the type of batch data
```

Comment:
We need to prevent the batch poster from posting a fake 4844 batch here by just specifying a header byte of `DATA_BLOB_HEADER_FLAG`. In addition, we need to prevent this from happening before the smart contract is upgraded. That's why the first 4844 PR used the `DATA_AUTHENTICATED_FLAG`

Reply:
Wouldnt that be prevented by the datahash reader erroring or returning no blobs?

Reply:
I think this would be prevented by looking at the batch data location in the SequencerBatchDelivered event.

Reply:
The replay binary doesn't have access to that information, only the batch accumulator

Reply:
Hmm, ok, in that case I think the first PR was vulnerable too. I think we should check the first byte in all seq inbox functions to ensure the correct type of data (normal, da, blob, etc) is being provided in each case.

Prior to 4844 going live the data hash reader will not be able to return any hashes so the blob seq inbox function will revert. My assumption for nitro is that it would support 4844 blobs from the moment the EIP goes live or from the point that the seq inbox is upgraded. So I think with the above additional check we should be ok right?

Reply:
The problem is that nitro doesn't know when the sequencer inbox contract is upgraded. If you're running a 4844 compatible nitro node (or replay binary) against an unupgraded sequencer inbox smart contract, and the batch poster maliciously posts a normal non-4844 batch with a first byte of `DATA_BLOB_HEADER_FLAG`, the replay binary doesn't have a good way of knowing that it's not a real 4844 batch and it shouldn't resolve the data hashes in it.